feat: EVM emulator docs (#333)
matter-labs/zksync-docs: content/00.zksync-era/30.unique-features/40.evm-emulator/01.evm-emulator.md

---
title: Overview
description: The special contracts that allows executing EVM bytecode on EraVM
---

> ZKsync Era integrates an EVM bytecode emulator that enables compatibility with EVM bytecode by translating EVM instructions into EraVM

Reviewer: Just a suggestion, this sentence feels a bit long.

```suggestion
ZKsync Era integrates an EVM bytecode emulator that enables compatibility with EVM bytecode. The emulator works by translating EVM instructions into EraVM
```

instructions at execution time.

This allows developers to deploy and interact with smart contracts without the need of using the ZKsync custom compilers

> (`zksolc` and `zkvyper`) Additionally, standard EVM tooling like Foundry and Hardhat works out of the box without the need for plugins.

Reviewer: ```suggestion
(`zksolc` and `zkvyper`). Additionally, standard EVM tooling like Foundry and Hardhat works out of the box without the need for plugins.
```

---
title: Overview
description: The special contracts that allows executing EVM bytecode on EraVM
---
ZKsync Era integrates an EVM emulator that enables compatibility with EVM bytecode. The emulator works by translating EVM instructions into EraVM
instructions at execution time.

This allows developers to deploy and interact with smart contracts without the need of using the ZKsync custom compilers
(`zksolc` and `zkvyper`). Additionally, standard EVM tooling like Foundry and Hardhat works out of the box without the need
for [plugins](../../tooling/hardhat/plugins/hardhat-zksync) or [custom versions](../../tooling/foundry/overview).
All while maintaining the core functionalities of EraVM, like [native account abstraction](../account-abstraction).

::callout{icon="i-heroicons-exclamation-triangle" color="amber"}
For a more detailed information about the EVM emulator, read the [protocol documentation](../../../zksync-protocol/contracts/evm-emulator/overview) and
the [Frequently Asked Questions](./faq).
::

## Main Features

> - **Standard Compiler Support**: Contracts do not need to be compiled with ZKsync-specific compilers (`zksolc` or `zkvyper`).

Reviewer 2: ```suggestion
- **EVM Bytecode Support**: Contracts do not need to be compiled with ZKsync-specific compilers (`zksolc` or `zkvyper`).
```

- **Tooling Compatibility**: Development tools such as Foundry, Hardhat, and Remix work without requiring additional plugins or modifications.
- **Address Derivation Consistency**: Addresses generated using `create` and `create2` match the behavior on the EVM.
- **Pre Deployed Contracts**: Provides contracts such as `create2` , `multicall3`  and
  `singletonFactory` (ERC2470). Refer to the [protocol documentation](../../../zksync-protocol/contracts/evm-emulator/pre-deployed-contracts) for

>   additional the full list.

Reviewer 2: ```suggestion
  the full list.
```

## Limitations

> - **Debugging**: EVM-like debugging is not available due to missing RPC methods.

Reviewer 2: ```suggestion
- **Debugging**: EVM-compatible debugging is not supported yet due to EraVM-specific nuances of the system.
```

- **Gas Behavior**: Gas calculations follow EraVM logic, not EVM gas logic.

> - **Unsupported Opcodes**: The same opcodes unsupported in EraVM remain unsupported. See the

Reviewer 2: Let's list unsupported opcodes here
